### PR TITLE
Remove paddle.where in cross_tropy_loss

### PR DESCRIPTION
**Pull request by HydrogenSulfate**
### PR types
Function optimization

### PR changes
APIs
### Describe
1. 使用已修复BUG的paddle.min和paddle.max代替paddle.where来检查labels合法性，减少GPU和CPU之间不必要的同步

### CODE REVIEW

**Review by pangyoki — COMMENTED** [review]
可以描述下删掉的这些check的功能和删掉之后有啥影响吗？为什么之前需要加上这个checking，现在直接删掉的时候，出现这些case是有其它代码会检查吗？

**Comment by HydrogenSulfate** [comment]
> 可以描述下删掉的这些check的功能和删掉之后有啥影响吗？为什么之前需要加上这个checking，现在直接删掉的时候，出现这些case是有其它代码会检查吗？

去掉这个检查后，如果labels范围超出边界，则会在gather_nd这里报错
```python
def test_LabelValue_ExceedMax():
                input_data = paddle.rand(shape=[20, 100])
                label_data = paddle.randint(
                    0, 100, shape=[20, 1], dtype="int64")  # hard label
                label_data[0] = 100
                weight_data = paddle.rand([100])  # provide weight
                paddle.nn.functional.cross_entropy(
                    input=input_data,
                    label=label_data,
                    weight=weight_data,
                    ignore_index=-100)


Error: /paddle/paddle/fluid/operators/gather.cu.h:62 Assertion `index_value >= 0 && index_value < input_dims[j]` failed. The index is out of bounds, please check whether the dimensions of index and input meet the requirements. It should be less than [100] and greater than or equal to 0, but received [0]
```

```python
def test_LabelValue_ExceedMin():
                input_data = paddle.rand(shape=[20, 100])
                label_data = paddle.randint(
                    0, 100, shape=[20, 1], dtype="int64")  # hard label
                label_data[0] = -1
                weight_data = paddle.rand([100])  # provide weight
                paddle.nn.functional.cross_entropy(
                    input=input_data,
                    label=label_data,
                    weight=weight_data,
                    ignore_index=-100)

Error: /paddle/paddle/fluid/operators/gather.cu.h:62 Assertion `index_value >= 0 && index_value < input_dims[j]` failed. The index is out of bounds, please check whether the dimensions of index and input meet the requirements. It should be less than [100] and greater than or equal to 0, but received [0]
```

**Comment by pangyoki** [comment]
> > 可以描述下删掉的这些check的功能和删掉之后有啥影响吗？为什么之前需要加上这个checking，现在直接删掉的时候，出现这些case是有其它代码会检查吗？
> 
> 去掉这个检查后，如果labels范围超出边界，则会在gather_nd这里报错
> 
> ```python
> def test_LabelValue_ExceedMax():
>                 input_data = paddle.rand(shape=[20, 100])
>                 label_data = paddle.randint(
>                     0, 100, shape=[20, 1], dtype="int64")  # hard label
>                 label_data[0] = 100
>                 weight_data = paddle.rand([100])  # provide weight
>                 paddle.nn.functional.cross_entropy(
>                     input=input_data,
>                     label=label_data,
>                     weight=weight_data,
>                     ignore_index=-100)
> 
> 
> Error: /paddle/paddle/fluid/operators/gather.cu.h:62 Assertion `index_value >= 0 && index_value < input_dims[j]` failed. The index is out of bounds, please check whether the dimensions of index and input meet the requirements. It should be less than [100] and greater than or equal to 0, but received [0]
> ```
> 
> ```python
> def test_LabelValue_ExceedMin():
>                 input_data = paddle.rand(shape=[20, 100])
>                 label_data = paddle.randint(
>                     0, 100, shape=[20, 1], dtype="int64")  # hard label
>                 label_data[0] = -1
>                 weight_data = paddle.rand([100])  # provide weight
>                 paddle.nn.functional.cross_entropy(
>                     input=input_data,
>                     label=label_data,
>                     weight=weight_data,
>                     ignore_index=-100)
> 
> Error: /paddle/paddle/fluid/operators/gather.cu.h:62 Assertion `index_value >= 0 && index_value < input_dims[j]` failed. The index is out of bounds, please check whether the dimensions of index and input meet the requirements. It should be less than [100] and greater than or equal to 0, but received [0]
> ```

这样是不是就不容易判断出是cross_entropy的报错了啊，有python栈可以看出是cross_entropy的问题吗？
静态图也是相同的报错吗？是应该把动态图位置的报错删掉，还是说应该为静态图加上报错呢？

**Comment by rgn9536903** [comment]
500000

File: python/paddle/nn/functional/loss.py [hunk 1]
```diff
@@ -1666,25 +1666,16 @@ def cross_entropy(input,
...
-            # TODO: Temporarily use paddle.nonzero instead of paddle.max 
-            # to detect and find out possible illegal label values
-            if len(paddle.nonzero(valid_label >= input.shape[axis])) > 0:
-                invalid_label = paddle.gather_nd(
-                    valid_label,
-                    paddle.nonzero(valid_label >= input.shape[axis]))
-                raise ValueError(
-                    "Target({}) is out of class_dimension's upper bound({})".
-                    format(invalid_label[0], input.shape[axis] - 1))
+            valid_label = paddle.cast(
```
Comment (pangyoki) [inline python/paddle/nn/functional/loss.py:1669]: 之前讨论过在`weight is not None`的条件下，才需要判断label的合法性。需要把检查移动到`weight is not None`中吗？

Comment (HydrogenSulfate) [reply python/paddle/nn/functional/loss.py:1669]: 由于`softmax_with_cross_entropy`中没有对labels进行充分的合法性检查，所以这里在hard label条件下就判断label的合法性，无论是否用了weight

**Review by pangyoki — APPROVED** [review]
LGTM

**Comment by chajchaj** [comment]
LGTM
